src/group_theory/divisible.lean

```diff
@@ -0,0 +1,385 @@
+Authors: Jujian Zhang
+-/
+import group_theory.subgroup.pointwise
+import group_theory.quotient_group
+
+/-!
+# Divisible Group and rootable group
+
+In this file, we define a divisible add monoid and a rootable monoid with some basic properties.
+
+## Main definition
+* An additive monoid `A` is said to be divisible by `α` iff for all `n ≠ 0 ∈ α` and `y ∈ A`, there
+  is an `x ∈ A` such that `n • x = y`. In this file, we adpot a constructive approach, i.e. we ask
+  for an explicit `div : A → α → A` function such that `div a 0 = 0` and `n • div a n = a` for all
+  `n ≠ 0 ∈ α`.
+* A monoid `A` is said to be rootable by `α` iff for all `n ≠ 0 ∈ α` and `y ∈ A`, there is an
+  `x ∈ A` such that `x^n = y`. In this file, we adopt a constructive approach, i.e. we ask for an
+  explicit `root : A → α → A` function such that `root a 0 = 1` and `(root a n)ⁿ = a` for all
+  `n ≠ 0 ∈ α`.
+
+## Main results
+For additive monoids and groups:
+* `add_monoid.divisible_by_of_smul_surj` : the constructive definition of divisiblity is implied by
+  the condition that `n • x = a` has solutions for all `n ≠ 0` and `a ∈ A`.
+* `add_monoid.smul_surj_of_divisible_by` : the constructive definition of divisiblity implies
+  the condition that `n • x = a` has solutions for all `n ≠ 0` and `a ∈ A`.
+* `add_monoid.divisible_by_prod` : `A × B` is divisible for any two divisible additive monoids.
+* `add_monoid.divisible_by_pi` : any product of divisble additive monoids is divisible.
+* `add_group.divisible_by_int_of_divisible_by_nat` : for additive groups, int divisiblity is implied
+  by nat divisiblity.
+* `add_group.divisible_by_nat_of_divisible_by_int` : for additive groups, nat divisiblity is implied
+  by int divisiblity.
+* `add_comm_group.divisible_of_smul_top_eq_top` : the constructive definition of divisiblity is
+  implied by the condition that `n • A = A` for all `n ≠ 0`.
+* `add_comm_group.smul_top_eq_top_of_divisible` : the constructive definition of divisiblity implies
+  the condition that `n • A = A` for all `n ≠ 0`.
+* `add_comm_group.divisible_of_char_zero` : any field of characteristic zero is divisible.
+* `add_comm_group.divisible_by_quotient` : 1uotient group of divisible group is divisible.
+* `add_comm_group.divisible_by_of_surj` : if `A` is divisible and `A →+ B` is surjective, then `B`
+  is divisible.
+
+and their multiplicative counterparts:
+* `monoid.rootable_by_of_pow_surj` : the constructive definition of rootablity is implied by the
+  condition that `xⁿ = y` has solutions for all `n ≠ 0` and `a ∈ A`.
+* `monoid.pow_surj_of_rootable_by` : the constructive definition of rootablity implies the
+  condition that `xⁿ = y` has solutions for all `n ≠ 0` and `a ∈ A`.
+* `monoid.rootable_by_prod` : any product of two rootable monoids is rootable.
+* `monoid.rootable_by_pi` : any product of rootable monoids is rootable.
+* `group.divisible_by_int_of_divisible_by_nat` : in groups, int divisibility is implied by nat
```

Comment:
There are still a few `divisible` in the multiplicative parts.

Reply:
fixed.